File: modules/webrtc-core/src/main/java/com/shterneregen/securelan/webrtc/service/RtcMediaDeviceService.java
package com.shterneregen.securelan.webrtc.service;

import com.shterneregen.securelan.webrtc.event.RtcVideoFrameEvent;

import java.util.List;
import java.util.function.Consumer;

public interface RtcMediaDeviceService extends AutoCloseable {
    List<RtcMediaDevice> audioCaptureDevices();

    List<RtcMediaDevice> videoCaptureDevices();

    String testAudioCaptureDevice(String deviceId);

    String testVideoCaptureDevice(String deviceId);

    CameraPreviewSession startVideoPreview(String deviceId, Consumer<RtcVideoFrameEvent> frameConsumer);

    interface CameraPreviewSession extends AutoCloseable {
        String statusMessage();

        @Override
        void close();
    }

    @Override
    default void close() {
    }
}

File: modules/webrtc-core/src/main/kotlin/com/shterneregen/securelan/webrtc/service/RtcMediaDeviceService.kt
package com.shterneregen.securelan.webrtc.service

import com.shterneregen.securelan.webrtc.event.RtcVideoFrameEvent
import java.util.function.Consumer

interface RtcMediaDeviceService : AutoCloseable {
    fun audioCaptureDevices(): List<RtcMediaDevice>

    fun videoCaptureDevices(): List<RtcMediaDevice>

    fun testAudioCaptureDevice(deviceId: String?): String

    fun testVideoCaptureDevice(deviceId: String?): String

    fun startVideoPreview(deviceId: String?, frameConsumer: Consumer<RtcVideoFrameEvent>): CameraPreviewSession

    interface CameraPreviewSession : AutoCloseable {
        fun statusMessage(): String

        override fun close()
    }

    override fun close() {
    }
}
